### PR TITLE
Minimal non-Cartesian MRI + example on FastMRI breast

### PR DESCRIPTION
**Pull request by Andrewwango**
Minimal physics wrapper for `mri-nufft` + example showing that it works on fastMRI breast radial acquisition data on CPU, CUDA and MPS.

Closes old PRs #1089, #730 which were stale and/or had some over-complications

Figure_1

### Checks to be done before submitting your PR

- [x] `python3 -m pytest deepinv/tests` runs successfully.
- [x] `black .` and `ruff check .` run successfully.
- [x] `make html` runs successfully (in the `docs/` directory).
- [x] Updated docstrings related to the changes (as applicable).
- [x] Added an entry to the [changelog.rst](https://github.com/deepinv/deepinv/blob/main/docs/source/changelog.rst).

### LLM policy

LLM usage is ok, but not PRs generated 100% by AI. See our [LLM policy](https://deepinv.org/contributing.html#llm-policy) Tick below as appropriate:

- [x] I did not use LLM tools to write the code
- [ ] I, a human, wrote code, and LLM tools helped me to write part of the code.
- [ ] An LLM tool wrote all of the code.
- [ ] An agent submitted the PR and wrote the description.

### CODE REVIEW

**Comment by paquiteau** [comment]
Hi @Andrewwango Great to see this revival ! Any reason not go with the deepinv wrapper we already provide ? (see for instance this example: https://mind-inria.github.io/mri-nufft/generated/autoexamples/extras/example_recon.html) 

Anyway, please let us know if you have any questions / hiccups happening from this integration

File: deepinv/physics/nufft.py [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+
+from deepinv.physics.mri import MultiCoilMRI
+from deepinv.utils.mixins import MRIMixin
+
+
+class NonCartesianMRI(MultiCoilMRI, MRIMixin):
+    """
+    Non-Cartesian (multi-coil) MRI via `mri-nufft`.
+
+    This physics wraps non-uniform FFT forward and adjoint operators provided by `mri-nufft`, and models non-Cartesian MRI sequences such as
```
Comment (matthieutrs) [inline deepinv/physics/nufft.py]:
```suggestion
    This physics wraps non-uniform FFT forward and adjoint operators provided by the `mri-nufft` [library](https://github.com/mind-inria/mri-nufft/tree/master), and models non-Cartesian MRI sequences such as
```

File: deepinv/physics/nufft.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+    radial or spiral sampling.
+
+    The physics also supports other `mri-nufft` functionality such as Voronoi density compensation.
```
Comment (matthieutrs) [inline deepinv/physics/nufft.py]:
```suggestion
    The physics also supports other `mri-nufft` functionality such as density compensation.
```

File: deepinv/physics/nufft.py [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    :param str density_mode: `None` (no compensation), `compensate` (multiply density in the adjoint), or `adjointness` (split as sqrt-density in both forward and adjoint so adjointness holds).
+        Set to `compensate` for adjoint or RSS.
+        For anything else (dagger, other recons), set to `None`.
+        Note that we do it in the wrapper rather that using `mri-nufft` internals.
+    :param torch.device device: physics device
+    """
+
+    def __init__(
+        self,
+        img_size: tuple[int, ...] = (320, 320),
```
Comment (matthieutrs) [inline deepinv/physics/nufft.py]:
```suggestion
        img_size: tuple[int, ...],
```
I think deepinv's convention is to not provide default image shapes, so I would avoid giving a default one

File: deepinv/physics/nufft.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+    We assume that `x` is of shape `(B,2,H,W)` and kspace `y` are `(B,2,N,S)` where `N` = coils and `S` = num shots * num samples per shot.
+
+    .. note::
+        Only supports 2D acquisition for now. For 3D/stacked physics, please open a feature request issue on GitHub.
+
+    .. warning::
+        The physics loops over batch elements, so will be slow for batch sizes > 1.
+
+    :param tuple img_size: reconstructed image size `(H, W)` (no channel dim), defaults to `(320, 320)` (fastMRI breast default).
```
Comment (matthieutrs) [inline deepinv/physics/nufft.py]:
```suggestion
    :param tuple img_size: reconstructed image size `(H, W)` (no channel dim), defaults to `(320, 320)` (fastMRI default).
```
iirc it's all of fastMRI, not just breast?

File: deepinv/physics/nufft.py [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+        else:
+            raise ValueError(
+                "`density_mode` must be None, 'compensate', or 'adjointness'"
+            )
+        self.register_buffer(
+            "density", density.to(device) if density is not None else None
+        )
+        self.density_mode = density_mode
+
+        self.operator_norm = 1.0
```
Comment (matthieutrs) [inline deepinv/physics/nufft.py:128]: this looks very dangerous - what if the norm is not 1? Is there a situation that someone obtains the info that norm is 1, although it's not?

I understand the purpose though - maybe a different name, e.g. `rescale_norm_factor`?

File: deepinv/physics/nufft.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+                    for yb in self.to_torch_complex(y).numpy(force=True)
+                ]
+            )
+        ).to(y.device)
+
+    def noise(self, x, **kwargs) -> torch.Tensor:
+        r"""
+        Bypass MultiCoilMRI Cartesian masked noise
+        """
+        return self.noise_model(x, **kwargs)
```
Comment (matthieutrs) [inline deepinv/physics/nufft.py:237]: if applying density compensation in an adjointness-compatible way, should there not be a rescale there?

Comment (matthieutrs) [reply deepinv/physics/nufft.py:237]: I think we need to add a test that simulating `density_mode='compensate'` `density_mode='adjoint'` the same vector `AtAy` (I'm not worried for the operator itself, but for the noise)

File: examples/external-libraries/demo_mrinufft_breast.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+# density compensation, which is standard for non-Cartesian MRI data.
+#
+# The FastMRI data was acquired with golden-angle radial sampling, with 640 samples per shot.
+# We use the standard reconstruction size of 320*320.
+
+physics_fs = dinv.physics.NonCartesianMRI(
+    img_size=(320, 320),
+    num_shots=288,
+    num_samples_per_shot=640,
+    coil_maps=y.shape[2],  # dummy coil maps, not used for RSS
```
Comment (matthieutrs) [inline examples/external-libraries/demo_mrinufft_breast.py]: maybe here the argument is not clear - I would expect coil maps, not a shape

File: examples/external-libraries/demo_mrinufft_breast.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+    noise_model=dinv.physics.GaussianNoise(0.001),
+)
+
+coil_maps = physics.estimate_coil_maps(y, method="espirit", decim=4)
+physics.update(coil_maps=coil_maps)
+
+# %%
+# Reconstruct with conjugate-gradient
+# -----------------------------------
+# We reconstruct the data with the conjugate-gradient algorithm, which gives a least-squares solution.
```
Comment (matthieutrs) [inline examples/external-libraries/demo_mrinufft_breast.py]:
```suggestion
# We reconstruct the data with the conjugate-gradient (CG) algorithm, which gives a least-squares solution.
```

**Comment by Andrewwango** [comment]
Hey @paquiteau , thanks for the reply! 

I'm aware of `make_deepinv_phy` and I think it makes the most sense to have both wrappers, to have bidirectional integration. Either way, both of them are very thin wrappers anyway, and all rely on the underlying mechanisms of mri-nufft (different backends, trajectories etc.). I will add some more comments pointing to the further functionality inside mri-nufft such that our wrapper functions as a signpost for users to explore mri-nufft more. Also, in future development of wrapper proposed here, I'll make sure we use underlying mri-nufft functionality as much as possible. 

Note that this is in line with deepinv philosophy (e.g. see our Tomography wrapper with Astra, or or PET wrapper with parallelproj) - we maintain abstracted physics which all meet our standards of documentation, conventions of tensor shapes, and public method signatures.

File: deepinv/physics/nufft.py (hunk 4 above)
Comment (Andrewwango) [reply deepinv/physics/nufft.py]: No, for brain/knee it's just the crop size, not the recon size

File: deepinv/physics/nufft.py (hunk 5 above)
Comment (Andrewwango) [reply deepinv/physics/nufft.py:128]: I see what you mean! But this is the existing convention we have already in PET etc. 

I'll add a comment

File: deepinv/physics/nufft.py (hunk 6 above)
Comment (Andrewwango) [reply deepinv/physics/nufft.py:237]: solved offline

File: deepinv/tests/test_physics.py [hunk 9]
```diff
@@ -1161,6 +1197,16 @@ def test_MRI_noise_domain(mri, mri_img_size, device, rng):
...
 
 
+def test_NonCartesianMRI_density_compensation(device):
+    physics, imsize, _, dtype = find_operator("NonCartesianMRI", device)
+    x = torch.randn(imsize, device=device, dtype=dtype).unsqueeze(0)
+    y = physics.A(x)
+
+    x_dc = physics.A_dagger(y, density_compensate=True)
+    assert x_dc.shape == x.shape
+    assert not torch.allclose(x_dc, physics.A_adjoint(y))
```
Comment (matthieutrs) [inline deepinv/tests/test_physics.py:1207]:
```suggestion
    assert torch.allclose(x_dc, physics.A_adjoint(y * physics.density))
```
maybe write an explicit check instead

File: examples/external-libraries/demo_mrinufft_breast.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+#
+# .. note::
+#     The target x is on wrong scale for metrics since it uses density compensation which doesn't preserve norm
+
+with torch.no_grad():
+    x_cg = physics.A_dagger(y)
+    x_adj = physics.A_adjoint(y)
+    x_dc = physics.A_dagger(y, density_compensate=True)
+
+dinv.utils.plot(
```
Comment (matthieutrs) [inline examples/external-libraries/demo_mrinufft_breast.py:162]: maybe add inset zoom?

File: examples/external-libraries/demo_mrinufft_breast.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+# - `Y` = number of samples per spoke = 640,
+# - `N` = num coils = 16,
+# - `P` = partitions = 83 (this will become part of slices = 192)
+#
+# The data was acquired with a Cartesian fully-sampled partition (slice) readout, and is stored in the frequency domain.
+# Therefore, preprocessing steps are needed to zero-fill the partition axis, then iFFT the partition axis, then take a 2D slice.
+#
+# For the demo, we perform these steps offline and provide on HuggingFace a sample slice available to download.
+# To reproduce this slice preprocessing, you can run the following code::
+#
```
Comment (matthieutrs) [inline examples/external-libraries/demo_mrinufft_breast.py:57]:
```suggestion
# 
# .. code-block:: python
```

(display is broken otherwise)

**Review by matthieutrs — CHANGES_REQUESTED** [review]
thanks, that's almost perfect, 3 minor points missing I think :)

File: deepinv/physics/nufft.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+
+        # Normalizing physics: default = don't normalize: divide by 1 in A and adjoint.
+        # if normalize=True, divide by empirically calculated operator norm such that
+        # resulting operator has norm 1.
+        self.register_buffer("operator_norm", torch.tensor(1.0, device=device))
+        if normalize:
+            self.operator_norm = self.compute_norm(
+                torch.randn(1, 2, *self.img_size[-2:], device=device),
+                squared=False,
+            )
```
Comment (romainvo) [inline deepinv/physics/nufft.py:138]: I think we should provide a `self.normalize` attribute so that anyone can check at runtime if the instanciated physics is a normalized one or not

File: deepinv/physics/nufft.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+
+        if dtype is not None:  # pragma: no cover
+            # finufft plan requires initialising with correct dtype
+            self.samples = self.samples.astype(dtype)
+
+        self.backend = backend
+
+        _, op = mrinufft.operators.base.FourierOperatorBase.interfaces[self.backend]
+        self.E = op(
```
Comment (romainvo) [inline deepinv/physics/nufft.py:111]: `self.E` for Engine? For readability I think we should choose a more explicit variable name

File: deepinv/physics/nufft.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+    :param tuple img_size: reconstructed image size `(H, W)` (no channel dim).
+    :param int num_shots: number of sampling shots `Nc` (e.g. spokes)
+    :param int num_samples_per_shot: number of samples per shot `Ns`
+    :param str trajectory: `radial` or `spiral`, passed to `mri-nufft`.
+    :param str, float tilt: radial spoke tilt for mrinufft radial trajectory. Or set to `golden`/`grasp` and `in_out=True` for fastMRI breast data.
+    :param bool in_out: if `True`, radial spokes span the full diameter (edge-to-edge through centre) instead of centre-out.
+    :param torch.Tensor, int, None coil_maps: complex coil sensitivity maps of shape `(H,W)`, `(N,H,W)` or `(B,N,H,W)`. `int` `N` simulates `N` birdcage maps (requires `sigpy`). `None` = single-coil (flat map).
+    :param str backend: mri-nufft backend. Use `finufft` for CPU, `cufinufft` for CUDA. Set to `mps` to use finufft on Apple MPS, which avoids a torch threading clash with `libomp`.
+    :param bool normalize: normalise by empirical norm
+    :param torch.device device: physics device
```
Comment (romainvo) [inline deepinv/physics/nufft.py:42]: Please describe what is the default value for each argument

File: deepinv/tests/test_physics.py [hunk 15]
```diff
@@ -578,6 +580,40 @@ def find_operator(name, device, imsize=None, get_physics_param=False):
...
+            "mrinufft",
+            reason="This test requires mri-nufft. Install with "
+            "`pip install mri-nufft[finufft]` (CPU/MPS) or `mri-nufft[cufinufft]` (GPU).",
+        )
+        if torch.device(device).type == "cuda":
+            backend = "cufinufft"
+        if torch.backends.mps.is_available():
+            backend = "mps"
+        else:
+            backend = "finufft"
```
Comment (romainvo) [inline deepinv/tests/test_physics.py:594]: `NonCartesianMRI` physics were always instanciated with the "finufft" backend, which caused additional warnings to be raised in `test_operators_norm` below

```suggestion
        if torch.device(device).type == "cuda":
            backend = "cufinufft"
        elif torch.backends.mps.is_available():
            backend = "mps"
        else:
            backend = "finufft"
```

File: deepinv/tests/test_physics.py [hunk 16]
```diff
@@ -918,7 +954,7 @@ def test_operators_norm(name, verbose, device, rng):
     with warnings.catch_warnings(record=True) as w:
         warnings.simplefilter("always")
         physics.compute_sqnorm(x, max_iter=1, tol=1e-9, verbose=verbose)
-        assert len(w) == 1
+        assert len(w) == 1 if name != "NonCartesianMRI" else 3
```
Comment (romainvo) [inline deepinv/tests/test_physics.py:957]: Problem solved above so no need for exception
```suggestion
        assert len(w) == 1
```

**Review by romainvo — CHANGES_REQUESTED** [review]
Thanks @Andrewwango for the very cool wrapper! Few remarks not too technical.

File: deepinv/tests/test_physics.py (hunk 15 above)
Comment (Andrewwango) [reply deepinv/tests/test_physics.py:594]: lol thanks Claude for this silly mistake, sorry

File: deepinv/tests/test_physics.py (hunk 16 above)
Comment (Andrewwango) [reply deepinv/tests/test_physics.py:957]: ahhhhh makes sense thanks for finding this!

**Review by matthieutrs — APPROVED** [review]

**Review by romainvo — APPROVED** [review]
Thanks @Andrewwango for the last fixes! Approving as failing tests are unrelated
